providers/apis/defi/raydium/provider.go

```diff
@@ -0,0 +1,253 @@
+// will be returned.
+func NewAPIPriceFetcherWithClient(
+	market oracletypes.ProviderMarketMap,
+	config config.APIConfig,
+	client SolanaJSONRPCClient,
+	logger *zap.Logger,
+) (*APIPriceFetcher, error) {
+	if err := config.ValidateBasic(); err != nil {
+		return nil, fmt.Errorf("config for raydium is invalid: %w", err)
+	}
+
+	if err := market.ValidateBasic(); err != nil {
+		return nil, fmt.Errorf("market config for raydium is invalid: %w", err)
+	}
+
+	// check fields of config
+	if config.Name != Name {
+		return nil, fmt.Errorf("config.Name is not %s", Name)
+	}
+
+	if market.Name != Name {
+		return nil, fmt.Errorf("market.Name is not %s", Name)
+	}
+
+	if !config.Enabled {
+		return nil, fmt.Errorf("config is not enabled")
+	}
+
+	// generate metadata per ticker
+	metadataPerTicker := make(map[string]TickerMetadata)
+	for _, ticker := range market.OffChainMap {
+		metadata, err := unmarshalMetadataJSON(ticker.Metadata_JSON)
+		if err != nil {
+			return nil, fmt.Errorf("error unmarshalling metadata for ticker %s: %w", ticker.String(), err)
+		}
+
+		if err := metadata.ValidateBasic(); err != nil {
+			return nil, fmt.Errorf("metadata for ticker %s is invalid: %w", ticker.String(), err)
+		}
+
+		metadataPerTicker[ticker.String()] = metadata
+	}
+
+	return &APIPriceFetcher{
+		market:            market,
+		config:            config,
+		client:            client,
+		metaDataPerTicker: metadataPerTicker,
+		logger:            logger.With(zap.String("provider", Name)),
+	}, nil
+}
+
+// FetchPrices fetches prices from the solana JSON-RPC API for the given currency-pairs. Specifically
+// for each currency-pair,
+//   - Query the raydium API base (coin) / quote (pc) token vault addresses
+//   - Normalize the token balances by 1e18
+//   - Calculate the price as quote / base, and scale by ticker.Decimals
+func (pf *APIPriceFetcher) Fetch(
+	ctx context.Context,
+	tickers []mmtypes.Ticker,
+) providertypes.GetResponse[mmtypes.Ticker, *big.Int] {
+	// get the acounts to query in order of the tickers given
+	expectedNumAccounts := len(tickers) * 2
+	accounts := make([]solana.PublicKey, expectedNumAccounts)
+
+	for i, ticker := range tickers {
+		metadata, ok := pf.metaDataPerTicker[ticker.String()]
+		if !ok {
+			return providertypes.NewGetResponseWithErr[mmtypes.Ticker, *big.Int](
+				tickers,
+				providertypes.NewErrorWithCode(
+					NoRaydiumMetadataForTickerError(ticker.String()),
+					providertypes.ErrorUnknownPair,
+				),
+			)
+		}
+
+		accounts[i*2] = solana.MustPublicKeyFromBase58(metadata.BaseTokenVault.TokenVaultAddress)
+		accounts[i*2+1] = solana.MustPublicKeyFromBase58(metadata.QuoteTokenVault.TokenVaultAddress)
+	}
+
+	// query the accounts
+	accountsResp, err := pf.client.GetMultipleAccountsWithOpts(ctx, accounts, &rpc.GetMultipleAccountsOpts{
+		Commitment: rpc.CommitmentFinalized,
+		// TODO(nikhil): Keep track of latest height queried as well?
+	})
```

Comment:
Are we certain that the resp is sorted in the same order we make the request?

Reply:
Is there a way for us to deterministically very this?

Reply:
The solana JSON-RPC server implementation orders the response in the order that accounts are requested. We don't get account addresses in response so we can't verify account data v account address

Reply:
mmm

Reply:
ok i would leave a godoc for that